contracts/exchange/contracts/test/TestTransactions.sol

```diff
@@ -0,0 +1,110 @@
+/*
+
+  Copyright 2019 ZeroEx Intl.
+
+  Licensed under the Apache License, Version 2.0 (the "License");
+  you may not use this file except in compliance with the License.
+  You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+  Unless required by applicable law or agreed to in writing, software
+  distributed under the License is distributed on an "AS IS" BASIS,
+  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  See the License for the specific language governing permissions and
+  limitations under the License.
+
+*/
+
+pragma solidity ^0.5.9;
+pragma experimental ABIEncoderV2;
+
+import "../src/Exchange.sol";
+import "../src/MixinTransactions.sol";
+
+
+contract TestTransactions is
+    Exchange
+{
+    // Indicates whether or not the overridden _isValidTransactionWithHashSignature should return true or false.
+    bool public shouldBeValid;
+
+    // Indicates whether or not the fallback function should succeed.
+    bool public shouldSucceedCall;
+
+    // The returndata of the fallback function.
+    bytes public fallbackReturnData;
+
+    constructor ()
+        public
+        Exchange(1337)
+    {} // solhint-disable-line no-empty-blocks
+
+    // This fallback function will succeed if the bool `shouldSucceedCall` has been set
+    // to true, and will fail otherwise. It will return returndata `fallbackReturnData`
+    // in either case.
+    function ()
```

Comment:
Hmm, I'm not sure if this is the clearest way to accomplish these tests and I'm not convinced that they test reentrancy intuitively. There's also some setup you need to do for each test, so I think the contract may be more complicated than it needs to be.

What if, instead, we had a function:

```solidity
function executable(bool shouldSucceed, bytes memory callData, bytes memory returnData)
    public returns (bytes memory)
{
    emit ExecutableCalled(...);
    require(shouldSucceed, 'EXECUTABLE_FAILED');
    if (callData.length != 0) {
        (bool didSucceed, bytes memory callResultData) = address(this).call(callData);
        if (!didSucceed) {
            assembly { revert(add(callResultData, 0x20), mload(callResultData)) }
        }
    }
    return bytes;
}
```

This way:
- The function's intent is clear.
- You get log entries to assert.
- You do not rely on any state to setup.
- You can easily test reentrancy (to any depth) by just encoding another
`executeTransaction()` call inside `callData`.

Does this make sense?

Reply:
This makes a lot of sense. I'll work on switching it over.